lib/mapObjects/MiscObjects.cpp

```diff
@@ -743,135 +744,217 @@ void CGResource::blockingDialogAnswered(const CGHeroInstance *hero, ui32 answer)
-		auto & tile = *RandomGeneratorUtil::nextItem(tiles, cb->gameState()->getRandomGenerator());
-		cb->moveHero(h->id, tile + int3(1,0,0), true);
-	}
+
+	cb->moveHero(h->id,CGHeroInstance::convertPosition(cb->getObj(destinationid)->pos,true) - getVisitableOffset(), true);
+}
+
+void CGMonolith::monolithDialogAnswered(const CGHeroInstance *hero, ui32 answer) const
+{
+	auto obj = cb->getObj(ObjectInstanceID(answer));
+	if(obj)
+		cb->moveHero(hero->id,CGHeroInstance::convertPosition(obj->pos,true) - getVisitableOffset(), true);
 	else
-		cb->moveHero (h->id,CGHeroInstance::convertPosition(cb->getObj(destinationid)->pos,true) - getVisitableOffset(), true);
+		throw std::runtime_error("Wrong teleporter");
 }
 
-void CGTeleport::initObj()
+void CGMonolith::initObj()
 {
-	int si = subID;
-	switch (ID)
+	std::vector<Obj> IDs;
+	IDs.push_back(ID);
+	switch(ID)
 	{
-	case Obj::SUBTERRANEAN_GATE://ignore subterranean gates subid
-	case Obj::WHIRLPOOL:
-		{
-			si = 0;
+		case Obj::MONOLITH_ONE_WAY_ENTRANCE:
+			type = ENTRANCE;
+			IDs.push_back(Obj::MONOLITH_ONE_WAY_EXIT);
+			break;
+		case Obj::MONOLITH_ONE_WAY_EXIT:
+			type = EXIT;
+			IDs.push_back(Obj::MONOLITH_ONE_WAY_ENTRANCE);
+			break;
+		case Obj::MONOLITH_TWO_WAY:
+		default:
+			type = BOTH;
 			break;
-		}
-	default:
-		break;
 	}
-	objs[ID][si].push_back(id);
+
+	channel = findMeChannel(IDs, subID);
+	if(!channel)
+	{
+		cb->gameState()->map->teleportChannels.push_back(TeleportChannel());
+		channel = &cb->gameState()->map->teleportChannels.back();
+	}
+	addToChannel();
+	objs.push_back(id);
 }
 
-void CGTeleport::postInit() //matches subterranean gates into pairs
+TeleportChannel * CGMonolith::findMeChannel(std::vector<Obj> IDs, int SubID) const
+{
+	for(auto objId : objs)
+	{
+		auto obj = dynamic_cast<const CGMonolith*>(cb->getObj(objId));
+		if(obj && vstd::contains(IDs, obj->ID) && obj->subID == SubID)
+			return obj->channel;
+	}
+
+	return nullptr;
+}
+
+void CGSubterraneanGate::onHeroVisit( const CGHeroInstance * h ) const
+{
+	ObjectInstanceID destinationid = getRandomExit();
+	if(destinationid == ObjectInstanceID()) //no exit
+	{
+		showInfoDialog(h,153,0);//Just inside the entrance you find a large pile of rubble blocking the tunnel. You leave discouraged.
+		logGlobal->warnStream() << "Cannot find exit subterranean gate for "<< id << " (obj at " << pos << ") :(";
+		return;
+	}
+
+	cb->moveHero(h->id,CGHeroInstance::convertPosition(cb->getObj(destinationid)->pos,true) - getVisitableOffset(), true);
+}
+
+void CGSubterraneanGate::initObj()
+{
+	objs.push_back(id);
+	type = BOTH;
+}
+
+void CGSubterraneanGate::postInit( CGameState * gs ) //matches subterranean gates into pairs
 {
 	//split on underground and surface gates
-	std::vector<const CGObjectInstance *> gatesSplit[2]; //surface and underground gates
-	for(auto & elem : objs[Obj::SUBTERRANEAN_GATE][0])
+	std::vector<CGMonolith *> gatesSplit[2]; //surface and underground gates
```

Comment:
I'm wish to put support for multi-level maps there, but I don't like idea about submitting any code I unable to test because it's may end up that it's have to be hugely rewritten later on anyway.

Though I already had idea to implement multiple channels per teleporter as it's would be first dependency of subterranean gates support the way we talked about. E.g when gates always lead to deeper level and one on deepest level lead to the top.